File: ArraysAndHashing/Java/3_TwoSum.java
/**
https://leetcode.com/problems/two-sum/

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
*/

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<int, int> track = new HashMap();
        for (int i = 0; i < nums.length; i++) {
            if (track.containsKey(target - nums[i])) return new int[] {i, track.get(target - nums[i])};
            track.put(nums[i], i);
        }       
        return null;
    }
}

/**
Algorithm: Maintain a hashset that keeps track of all the elements and their index. Then when visiting each number, check if the target - curr element is in the dict. If it is then we found the pair of elements that add to target

Space: O(n), all numbers and their indeces stored in the hashmap
Time: O(n), every num in the array visited
*/

File: ArraysAndHashing/C#/3_TwoSum.cs
/**
https://leetcode.com/problems/two-sum/

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
*/

public int[] TwoSum(int[] nums, int target) {
    Dictionary<int, int> track = new Dictionary<int, int>();
    for (int i = 0; i < nums.Length; i++) {
        if (track.ContainsKey(target - nums[i])) {
            return new int[2] {i, track[target - nums[i]]};
        }
        if (!track.ContainsKey(nums[i])) {
            track.Add(nums[i], i);
        }
    }
    return null;
}

/**
Algorithm:
Store each element in the dictionary. Check if target - element is already in the dictionary. If it is then we found our pair and return those indeces

Space: O(n), each element is stored in the dictionary
Time: O(n), at worst every element is visited
*/